WIP: #5605 fix Chinese search index

TimKam opened the pull request:
Subject: generate search index for Latin words correctly if search language is Chinese

### Feature or Bugfix
- Bugfix

### Purpose
- As described at #5605
- Conf: ``html_search_language = 'zh'``

### Detail
- Removes trailing (and preceding) spaces from Latin search index terms
- WIP, because a Chinese-speaking person needs to confirm the correctness

### Relates
- Fixes #5605

TimKam committed:
#5605 fix Chinese search index

generate search index for Latin words correctly if search language is Chinese

TimKam committed:
#5605 fix issue with stemmer, zh

TimKam committed:
#5605 add type annotation

TimKam commented:
@DaGeger can you test once again? What do you think, @tk0miya?

tk0miya commented on sphinx/search/zh.py:
```diff
@@ -250,13 +251,24 @@ def split(self, input):
+        # Don't stem Latin words that are long enough to be relevant for search
+        # if not stemmed, but would be too short after being stemmed
+        # avoids some issues with acronyms
+        should_not_be_stemmed =\
```
I feel folding with `\` is a bit ugly. How about using parenthesis like this?
```
should_not_be_stemmed = (
            word in self.latin1 and
            len(word) >= 3 and
            len(self.stemmer.stem(word.lower())) < 3
)
```

tk0miya reviewed (approved):
LGTM with nits

tk0miya reviewed (changes requested):
I noticed this should be merged into 1.8 branch because this is a bug fix.

TimKam committed:
#5605 improve code style

tk0miya reviewed (approved):
LGTM

tk0miya commented:
Is this still WIP?

TimKam commented:
Sorry, yes. I wrote some tests locally and found out it needs more work. Will get back to it at the weekend.

tk0miya commented:
Okay. Please let me know if you finished. I'll review this again :-)

TimKam commented:
@tk0miya Could you have another look? I needed to add a list that contains all Latin terms and change the regexp that splits Chinese from Latin characters. @DaGeger comments from you regarding the initial problem are also welcome.

tk0miya reviewed (approved):
LGTM!

TimKam committed:
#5605 add test assertion improve regexp

TimKam committed:
#5605 split chinese and Latin words correctly

And maintain list of latin terms that can be accessed to make decision if word
should be stemmend

TimKam committed:
#5605 add change note

TimKam committed:
Merge remote-tracking branch 'upstream/1.8' into 5605-fix-english-words-chinese-search

TimKam committed:
#5605 improve regexp

TimKam committed:
Merge remote-tracking branch 'upstream/1.8' into 5605-fix-english-words-chinese-search

TimKam committed:
#5605 move change note to 1.8.4
